testcases.py

```diff
@@ -996,6 +996,116 @@ def check(self) -> TestResult:
         return TestResult.FAILED
 
 
+class TestCaseNAT(TestCaseTransfer):
+    @staticmethod
+    def name():
+        return "nat"
+
+    @staticmethod
+    def abbreviation():
+        return "BN"
+
+    @staticmethod
+    def testname(p: Perspective):
+        return "transfer"
+
+    @staticmethod
+    def desc():
+        return (
+            "Transfer completes under frequent NAT port rebindings on the client side."
+        )
+
+    def get_paths(self):
+        self._files = [
+            self._generate_random_file(10 * MB),
+        ]
+        return self._files
+
+    @staticmethod
+    def scenario() -> str:
+        """ Scenario for the ns3 simulator """
+        return "nat --delay=15ms --bandwidth=10Mbps --queue=25 --first_rebind=1s --rebind_freq=3s"
+
+    def check(self) -> TestResult:
+        result = super(TestCaseNAT, self).check()
+        if result != TestResult.SUCCEEDED:
+            return result
+
+        tr_server = self._server_trace()._get_packets(
+            self._server_trace()._get_direction_filter(Direction.FROM_SERVER) + " quic"
+        )
+
+        ports = list(set(getattr(p["udp"], "dstport") for p in tr_server))
+
+        logging.info("Server saw these client ports: %s", ports)
+        if len(ports) > 1:
+            return TestResult.SUCCEEDED
+
+        logging.info("Server saw only a single client ports in use; test broken?")
+        return TestResult.FAILED
+
+
+class TestCaseCGN(TestCaseNAT):
+    @staticmethod
+    def name():
+        return "cgn"
+
+    @staticmethod
+    def abbreviation():
+        return "BC"
+
+    @staticmethod
+    def desc():
+        return 'Transfer completes under frequent "carrier-grade NAT" IP and port rebindings on the client side.'
+
+    @staticmethod
+    def scenario() -> str:
+        """ Scenario for the ns3 simulator """
+        return super(TestCaseCGN, TestCaseCGN).scenario() + " --cgn"
+
+    def check(self) -> TestResult:
+        if not self._keylog_file():
+            logging.info("Can't check test result. SSLKEYLOG required.")
+            return TestResult.UNSUPPORTED
+
+        result = super(TestCaseCGN, self).check()
+        if result != TestResult.SUCCEEDED:
+            return result
+
+        tr_server = self._server_trace()._get_packets(
+            self._server_trace()._get_direction_filter(Direction.FROM_SERVER) + " quic"
+        )
+
+        ips = list(set(getattr(p["ip"], "dst") for p in tr_server))
+
+        logging.info("Server saw these client addresses: %s", ips)
+        if len(ips) <= 1:
```

Comment:
Use a different failure `if len(ips) == 0`.
```suggestion
        if len(ips) == 1:
```

Reply:
That wouldn't catch the case where the server for some reason saw no client packets though.